### ISSUE TITLE
[Question] How does EBF sort features for bundling?

### ISSUE THREAD

**Issue by JoshuaC3** [microsoft/LightGBM#4221]
I am trying to understand how EFB sorts the features ready for bundling. It is Part 1 of EFB in this [blog](https://towardsdatascience.com/what-makes-lightgbm-lightning-fast-a27cf0d9785e).  
  

![image](https://user-images.githubusercontent.com/11645712/115868652-aea16700-a434-11eb-8230-c25de9f1ce74.png)

The part I don't understand is `G.sortByDegree` in the paper - what is degree?

Alternatively, I dont understand the phrase "Sort the features by count of non zero instances in descending order." from the blog.

Any intuition or example would be greatly appreciated.

**Comment by shiyu1994** [comment]
Hi @JoshuaC3. In the graph `G`, each node represents a feature, and each edge between two nodes means there're conflicts between the two corresponding features. So the degree of a node means the number of features which have conflicts with the corresponding feature of the node.  (Or the total number of conflicts with other features, if we count the number of conflicts as weights of edges when calculating the `degree`).
However, note that this sorting is expensive since calculating these degrees would require `O(#data x #feature x #feature)`, which is expensive. So as stated in the second paragraph in page 6 of the paper, we would instead order the features by their counts of non-missing values (since EFB is targeted for handling sparse features). This is exactly what the blog says.

**Comment by skullknight15** [comment]
What is the computational gain for histogram building using both GOSS and EFB ? Do we go from O(|DATA|* NB_features) to O (|AUB|* NB_bundles) where A nd B are the high and small gradient samples

**Comment by jameslamb** [comment]
thanks for your interest in LightGBM @skullknight15 . Tagging reviewers who can help, since they might not see this comment on a closed issue.

@shiyu1994 @tongwu-msft @hzy46

**Comment by shiyu1994** [comment]
Hi @skullknight15, yes the computational gain of GOSS is reduce from O(|DATA|* NB_features) to O (|AUB|* NB_bundles) as you mentioned.
The ideal computational gain for EFB is from O(|DATA|* NB_features) to O(number of non-missing values of all features), when all the sparse features is combined into fully dense features.

**Comment by skullknight15** [comment]
Just to confirm @shiyu1994, the algorithm starts first by applying EFB then GOSS or GOSS first then EFB?

**Comment by shiyu1994** [comment]
@skullknight15 EFB is a preprocessing stage which happens before training the first tree. And GOSS happens every iteration.

**Comment by skullknight15** [comment]
@shiyu1994 okay, so EFB is before training in order to reduce the feature space, then GOSS is used at every iteration, specifically GOSS is enabled at 20-th iteration as the doc states.
Thank you very much @shiyu1994 for your time, I think that you can close this issue.

**Comment by RunzhouZhao** [comment]
@shiyu1994 Hi, I can't understand why features needed to be sorted, because we need to go through all the features to complete the Greedy bundling, whether we sort features or not, we still need go through all the features.

**Comment by shiyu1994** [comment]
@RunzhouZhao The Greedy Bundling Algorithm imitates a greedy coloring algorithm. According to https://en.wikipedia.org/wiki/Greedy_coloring, starting from vertices with higher degree is a strategy used by greedy coloring very commonly.
According to https://watermark.silverchair.com/100085.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAAsQwggLABgkqhkiG9w0BBwagggKxMIICrQIBADCCAqYGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMcLQZr1cH0Ip1BDfvAgEQgIICd6uTnrFp0feeReh3ed6SeTSYRHl2tHytrlHrPZ9mhrku_kWaTJXOYpWCOoFMvf_MMCHdYyayhAQ5pBy6FnjZnGq_djoXfItRg9Dij4vdgfDkBt2-jjLbP5BTj2UleVl4qMKTsFEbl-HymqXs5HbDeMggOyrsBCLEmvIw5LiA8RH_UYPHeCSfww_6qdDPhoPZXRMNrrUzQVkPE_eqNEuEbIGsETUiI0gknVtwe6UOFTHuLUG4U7nLlbRwBX9wIIVvblN5AKKLlYUFTVjqtUOXQv1-s2QAfZ9yY48FCmXwvz8oWh7y-DHlHl1qlkJ_Ml6v1LpU18EQdjfN8-3msxFfDt32-M8AK0dXZcKhCUrGYupBYBH5EbK67Q3RdKusCuy1ETtWwe-xde1g9vzK2QlV-DaSL-G52PFPsPS_cw9RYUzI8-uaq1fDNh4BCyXp6djmlulNRelCG1Qg90G9rJKmxP7JSNiih-LBYfjpLT9IHaU7ufx8QiOjuUHR5qDsaYbG4JrHMOmUBEBbibrhw9DhDVuXQG145QNrSr4536_-NhhHdP-31uDua7SUsWOC2H7odPKcWveZL7ITpNiAUR_Zn4lRmxnVYCS4Hm6nPVjEPJrznJEYyXxXpDp-Me8oXEPaYRK67Yv1HmfbreIbHb1nHkvfIBKiKxtuUfbcSpDP2qKLXaMz12V27TDpY_dzByOa_Eb1Q_u5DPTHVzpawEDKMXLKF2cCmiXtRVFdnn7jMZMv2DEzhUttIlJsmObEzaOpCEkdYhgcU-ZSYnLtvOoApUXTpJR7-oM9f4c3jZVZsSXJwBVZ4B-CyHb2C8sKF2N6OEiwGkHnoSY
the greedy coloring algorithm deployed by Greedy Bundling (greedy coloring according to descending ordering of vertex degrees) has an upper bound for the total number of colors to use.

**Comment by RunzhouZhao** [comment]
@shiyu1994 
Hello, one more question, I want to know if the parameter zero_as_missing=false is effective for EFB algorithm and "ignoring zero features" in the 4th paragraph on page 6 in the paper. Sometime zero features are available, for example, field: gender has two kinds of values: 1 represents male and 0 represents female. If we ignore the 0(female) when we build the histogram, this field will be not work, and EFB is also the similar. 
And if we ignore the 0 features when we train the model, how we handle the 0 features in test data?

**Comment by shiyu1994** [comment]
@RunzhouZhao Thanks for the question. Please not that `EFB` is just for efficiency. It should not affect the syntax of the tree growing algorithm if there's no conflicts between groups. Thus, the `zero` bin (actually, in the code we use the `most frequent bin` instead of `zero` for more efficiency improvement), is not really ignored when split finding. Instead, it is ignored only when histogram construction. The bins ignored during histogram construction, will be recovered in subsequent best split finding, see
https://github.com/microsoft/LightGBM/blob/d4cdbcfe1b1db2d48e95665a8b358028c496cdb2/src/treelearner/serial_tree_learner.cpp#L404

**Comment by RunzhouZhao** [comment]
@shiyu1994 Thanks so much, the answer is very helpful. I understand the section of EFB. But if we ignore 0 bin when build the histogram, how we find the best split for the feature including 0 bin? I guess that we handle 0 bin same as null: put 0 bin into left branch and right branch and compute the gains respectively, and select the branch having larger gain, right? So if the parameter zero_as_missing = true, we ignore zero when building histogram, else we use zero bin when building histogram, that's right?

**Comment by shiyu1994** [comment]
> I guess that we handle 0 bin same as null: put 0 bin into left branch and right branch and compute the gains respectively, and select the branch having larger gain, right?

That only happens when `zero_as_missing=true`. Otherwise, 0 bin will be considered as a normal bin (Though it may not be included in the histogram if 0 bin is the most frequent bin. But we can RECOVRE the statistics in 0 bin easily given statistics in all other bins. But that's too technical detailed to discuss here.) in best split finding.

> So if the parameter zero_as_missing = true, we ignore zero when building histogram, else we use zero bin when building histogram

Regardless of `zero_as_missing`, as long as 0 bin is the most frequent bin, it will be ignored in histogram building.

**Comment by peiji1981** [comment]
> > I guess that we handle 0 bin same as null: put 0 bin into left branch and right branch and compute the gains respectively, and select the branch having larger gain, right?
> 
> That only happens when `zero_as_missing=true`. Otherwise, 0 bin will be considered as a normal bin (Though it may not be included in the histogram if 0 bin is the most frequent bin. But we can RECOVRE the statistics in 0 bin easily given statistics in all other bins. But that's too technical detailed to discuss here.) in best split finding.
> 
> > So if the parameter zero_as_missing = true, we ignore zero when building histogram, else we use zero bin when building histogram
> 
> Regardless of `zero_as_missing`, as long as 0 bin is the most frequent bin, it will be ignored in histogram building.

hi, shiyu. Can you share some annotation of EFB related codes in Lightgbm?
